### ISSUE TITLE
uv sync fails on older uv versions

### ISSUE THREAD

**Issue by vmirly** [natolambert/rlhf-book#366]
Hi Nathan, thanks again for all the work you’ve put into the book. I’ve really enjoyed working through it.

I ran into an installation issue while setting up the code environment for the repo, so I wanted to share it in case it helps others

When running:
uv sync
I got this warning/error:
```
warning: Failed to parse `pyproject.toml` during settings discovery:
  TOML parse error at line 60, column 10
     |
  60 | [tool.uv.extra-build-dependencies]
     |          ^^^^^^^^^^^^^^^^^^^^^^^^
  unknown field `extra-build-dependencies`, expected one of `required-version`, `native-tls`, `offline`, `no-cache`, `cache-dir`, `preview`, `python-preference`, `python-downloads`, `concurrent-downloads`, `concurrent-builds`, `concurrent-installs`, `index`, `index-url`, `extra-index-url`, `no-index`, `find-links`, `index-strategy`, `keyring-provider`, `allow-insecure-host`, `resolution`, `prerelease`, `fork-strategy`, `dependency-metadata`, `config-settings`, `no-build-isolation`, `no-build-isolation-package`, `exclude-newer`, `link-mode`, `compile-bytecode`, `no-sources`, `upgrade`, `upgrade-package`, `reinstall`, `reinstall-package`, `no-build`, `no-build-package`, `no-binary`, `no-binary-package`, `python-install-mirror`, `pypy-install-mirror`, `python-downloads-json-url`, `publish-url`, `trusted-publishing`, `check-url`, `pip`, `cache-keys`, `override-dependencies`, `constraint-dependencies`, `build-constraint-dependencies`, `environments`, `required-environments`, `conflicts`, `workspace`, `sources`, `managed`, `package`, `default-groups`, `dev-dependencies`, `build-backend`

Using CPython 3.12.3 interpreter at: /usr/bin/python3
Creating virtual environment at: .venv
Resolved 133 packages in 0.56ms
  × Failed to build `pycosat==0.6.6`
  ├─▶ The build backend returned an error
  ╰─▶ Call to `setuptools.build_meta:__legacy__.build_wheel` failed (exit status: 1)

      [stderr]
      error: command 'x86_64-linux-gnu-gcc' failed: No such file or directory

      hint: This usually indicates a problem with the package or the build environment.
  help: `pycosat` (v0.6.6) was included because `rlhf-book-code` (v0.2.0) depends on `reasoning-gym` (v0.1.25) which depends on `pycosat`
```

There were actually two separate issues on my machine:
1. My installed version of uv was too old and did not recognize:
`[tool.uv.extra-build-dependencies]`
2. My Ubuntu system was missing the compiler toolchain needed to build pycosat (gcc was not installed).

What fixed it for me on Ubuntu 24.04 was:
```
sudo apt update
sudo apt install -y build-essential python3-dev
uv self update
```
Then I removed the old virtual environment and re-ran:
```
rm -rf .venv
uv sync
```
After that, the install worked.

**Comment by casinca** [comment]
Yes same, I had my venv ~~break~~ throw a warning at some point between 2 recent merges, so I guess it was coming from this PR https://github.com/natolambert/rlhf-book/pull/351

**Comment by natolambert** [comment]
What hardware is everyone on? Thanks @vmirly and @casinca 
I'm going to rent some other machines (or I could buy too) to make sure this works.

Also curious how widely @zafstojano has tested the code (ie how many GPU types)

**Comment by zafstojano** [comment]
Regarding Flash Attn, in my original repo I [found](https://github.com/zafstojano/policy-gradients?tab=readme-ov-file#getting-started) installing from pre-built wheels based on the specific versions of Python, CUDA and Torch worked best, e.g. 
```bash
uv pip install https://github.com/mjun0812/flash-attention-prebuild-wheels/releases/download/v0.4.17/flash_attn-2.8.3+cu128torch2.9-cp312-cp312-linux_x86_64.whl
```

When building the repo I had access to a 4x A10G machine (with the above dependencies)

Regarding the Reasoning Gym dependency on Ubuntu machines, can confirm that `build-essential` is needed (as seen in [this PR](https://github.com/meta-pytorch/OpenEnv/pull/326#issuecomment-3896231255) when I contributed it to Meta's OpenEnv)

**Comment by casinca** [comment]
> What hardware is everyone on? Thanks [@vmirly](https://github.com/vmirly) and [@casinca](https://github.com/casinca) I'm going to rent some other machines (or I could buy too) to make sure this works.
> 
> Also curious how widely [@zafstojano](https://github.com/zafstojano) has tested the code (ie how many GPU types)

Sorry I should have been more accurate, I was only talking about the first issue being the warning. But I re-tested across 3 OS windows(cuda)/mac(default to cpu)/linux(A100) it didn't "break" it's just the warning but training works.

I confirm the warning, in my case, was introduced by https://github.com/natolambert/rlhf-book/pull/351 if I recreate a venv from the branch before 351 (`git switch -c pre-fa-pr 79f4a5d`), I don't get the warning anymore

**Comment by natolambert** [comment]
Closed this with some documentation. Folks can re-open if not good enough or new issues emerge.
